numpy/_core/src/multiarray/datetime.c

```diff
@@ -2122,8 +2125,7 @@ NpyDatetime_ConvertPyDateTimeToDatetimeStruct(
     }
 
     /* Get the hour */
-    tmp = PyObject_GetAttrString(obj, "hour");
-    if (tmp == NULL) {
+    if (PyObject_GetOptionalAttr(obj, npy_interned_str.hour, &tmp) == -1) {
```

Comment:
This still isn't right; if this returns 0, then you need to return the date-only version. Continuing with `tmp` not set as you currently do is a bug.

Perhaps it would be best to write some tests before going further; to be able to see the impact here, you need to make some custom objects, which:
* Throw errors when `.hour` is accessed
* Succeed the first time that `.hour` is accessed, but fails the second time.

Reply:
hmm, what i thought is PyObject_HasAttrString will be good to check for the 0 cases ?

`if (!PyObject_HasAttrString(obj, "hour") ||
            !PyObject_HasAttrString(obj, "minute") ||
            !PyObject_HasAttrString(obj, "second") ||
            !PyObject_HasAttrString(obj, "microsecond")) {
        /* The best unit for date is 'D' */
        if (out_bestunit != NULL) {
            *out_bestunit = NPY_FR_D;
        }
        return 0;
    }`
    
    I mean here if the attr does not exist, it will handle the 0 case ? 
    
    -1 we are handling with : if (PyObject_GetOptionalAttr(obj, npy_interned_str.hour, &tmp) == -1)
    
    and 1 is good to go with, 
    
    No ?

Reply:
It is possible for `HasAttr` to return true, but for the attribute to not be present when you ask a second time:
```python
class OhNo:
    _count = 0
    @property
    def date(self):
        self._count += 1
        if self._count % 2 == 0:
            raise AttributeError
        else:
            return "today"
            
o = OhNo()
print(hasattr(o, "date"))
print(o.date) -- fails
```